### ISSUE TITLE
People stuck trying to enter airlock

### ISSUE THREAD

**Issue by bevans2000** [mars-sim/mars-sim#305]
**Describe the bug**
People are getting stuck entering an airlock from the inside. They are waiting for the airlock to become usable. This shows itself by a Task repeatedly not consuming any time.

**To Reproduce**
Not easy to reproduce.  But I think the problem occurs if an airlock is used a second time where the outer door is not shut by the previous person?.

**Additional context**
Add a command to the Console to show the status of the airlocks maybe including the last user. This may help track the problem.

**Comment by bevans2000** [comment]
So there are 2 problems going on here

1.  The new approach to the clock is very sensitive to Tasks that do not consume time. When using an Airlock people have to wait for them to cycle and that is causing some false negatives. The `Airlock.addTime` method now reports how much time has been consumed so the Task consumes time when waiting.
2. There is a situation where a Person is outside an Airlock requestingEgress but the Airlock is DEPRESSURISED with the Outer door unlocked. In this situation then Person gets stuck at the Task with nothing progressing. I don't know how the Airlock resets after someone has used it.

The `Mind.takeAction` will be changed to compensate for steps that do not consume time. A count of consecutive executions that do not consume time is used instead to make sure the method does not go into an infinite loop. There is no loner a control per Mind as some Tasks can take a long time. This change will be better to adapt to the Tasks.
Generally the executeTask method should consume time even if they are waiting.

**Comment by mokun** [comment]
@bevans2000 , you may have already figured this out. But to aid others in debugging this walking bug, I'd post the two rough drafts showing the definition of the zones :

1. EnterAirlock (used when entering an airlock from outside)

![IMG_20201230_120437](https://user-images.githubusercontent.com/1168584/103378707-169ffc00-4a98-11eb-9fe2-3caa37a0e8c1.jpg)

2. ExitAirlock (used when exiting an airlock from inside)

![IMG_20201230_120208](https://user-images.githubusercontent.com/1168584/103378751-43541380-4a98-11eb-9b70-dfda32f809fd.jpg)

**Comment by bevans2000** [comment]
What makes an Airlock re-pressurise after it has been used? I found cases where the Person is in Z0 in a ExitAirlock Task calling the requestEgress method but the outer door is unlocked and it is depressurised. The Person just spins around the method each clock pulse because nothing is changing. Sometimes someone else uses the Airlock to enter from the Outside and that seems to reset it so the waiting Person can use it.

**Comment by mokun** [comment]
@bevans2000 

> What makes an Airlock re-pressurise after it has been used?

**Comment by mokun** [comment]
@bevans2000 ,

I just updated the [Airlock](https://github.com/mars-sim/mars-sim/wiki/Airlock) wiki page.

**Comment by bevans2000** [comment]
@mokun  Where to start with this problem..... I have found 3 real issues:

1. There are ClassCastExceptions when people try to enter a Vehicle Airlock. This is because the logic using zones is only applicable to Building Airlocks but it is referenced in the generic flow. Method `Airlock.isInZone` is never going to work.
2. People get stuck in the enter airlock phase because the doors are not in the correct setting and they have past the opportunity to pressurise/depressurise. I think this is because someone else uses the airlock concurrently at the same time.
3. People get stuck trying to reach the innerdoor of the airlock because they are not in the correct position. A subtask is created but later in the requestEgress method the task is ended so they never reach the door

In a nutshell I don't know what to do with this. The split logic for Vehicle & Building airlocks makes the code flow very difficult to flow. We need to consolidate the approach, either both airlocks uses zones OR push the zone logic into the `BuildingAirlock` subclass and take it out of the Tasks.
The `EnterAirlock` & `ExitAirlock` have a lot of duplicated code which means the same "potential bugs" are in 2 places. There should be a common super class (e.g. `UseAirlock`) that has the common logic such as EVASuit, pressurizing. The subclasses should then just be responsible for the "direction" of travel through the phases.

I ran a simulation for ~10 Mars days and over 15 people where stuck at airlocks.

**Comment by mokun** [comment]
> @mokun Where to start with this problem..... I have found 3 real issues:

There is a few walking related classes. I'd say they are the main culprits.

> The `EnterAirlock` & `ExitAirlock` have a lot of duplicated code which means the same "potential bugs" are in 2 places. There should be a common super class (e.g. `UseAirlock`)

Yes. We can have a common super class if it can simplify things.

**Comment by bevans2000** [comment]
But what about zones? Should they be used in Vehicle Airlocks since it is a big hole at the moment.

**Comment by scud1** [comment]
I'm currently doing some work to fix bugs in the walking tasks and airlock code.  I'm working with a checkout of the code from back in October/November, so I maybe a bit out of date and have some merging work ahead of me.

I think I have the walking tasks fully functional again.  There were a few major bugs, including one that made people walk backwards when over encumbered.  I'd like to refactor the code a bit to clean it up.  I'm creating a Walkable interface that both Person and Robot classes will implement and hopefully prevent duplicate code in the Walk tasks.  Duplicate code is where a lot of bugs tend to crop up.

The airlock code, unfortunately, is very convoluted at this point and I think needs to be redone from scratch.  The original airlock code I wrote was poorly designed and has just gotten worse as new features have been added.  I had great difficulty following its behavior when stepping through the code in Eclipse.  So I'm currently working on redesigning the building and vehicle airlock code along with the EnterAirlock and ExitAirlock tasks.  I'm trying to keep the airlocks only as complex as necessary and making a clean separation between the active behavior and decision making of the humans/robots, and the passive/automatic behavior of the airlock.

Anyway, I'll let you know my progress with refactoring the airlock code.  Please let me know if you have any specific questions or suggestions related to it.

Scott

**Comment by bevans2000** [comment]
I like the sound of a Walkable interface. I thought of something simliar to represent a worker in  #297. Once you deliver this change we can use it as a base to do a refactoring of the Task area.

**Comment by mokun** [comment]
> But what about zones? 

I'd say defining zones are good because in each zone things happen uniquely and differently.

In future, we may even define in more detail than now, the real logistical things that need to happen in each zone for ingress and egress.

 > Should they be used in Vehicle Airlocks since it is a big hole at the moment.

I wonder if logistically speaking, we would treat the vehicular airlock the same way as the building airlock.

After we make everything working, we could fine tune what should happen in the vehicle. It should have less logistical tasks to go through than a full-blown building airlock.

**Comment by mokun** [comment]
@scud1 

> The airlock code, unfortunately, is very convoluted at this point and I think needs to be redone from scratch. The original airlock code I wrote was poorly designed and has just gotten worse as new features have been added. I had great difficulty following its behavior when stepping through the code in Eclipse. 

Scott, we've since redefined all the phases in the last 3 months or so. Though they bear the resemblance, the original airlock code have been heavily modified. 

I'm still wondering how well things work if a person already have a main task and he needs to start a subtask for walking to a destination.

In that case, `Mind` should be able to stop the main task and work on the walking subtask until it reaches the location. 

But what if for whatever reasons, he cannot arrive at the destination ? 

What would happen to the `Mind` class ? 

How should we design it to quit this bad walking sub-task ?

**Comment by scud1** [comment]
I'm still trying to understand your concept of zones and the 4 chambers of
an airlock.  I see them mentioned in several places but I'm not sure what
your overall concept is and how it would work.  Could you explain it a
little more clearly for me?

The airlock concept I'm working on is a fairly simple idea, although I'm
sure more could be added to it.  It's a small chamber built into a building
or vehicle with an inner and outer door.  It's a simple state machine with
two end states: "pressurized" and "depressurized", and two transitional
states: "pressurizing" and "depressurizing".  There is an activation button
both inside the airlock chamber and outside on a control panel next to each
door.  When the airlock is at an end state and activated, the airlock goes
into the appropriate transition state for the length of its cycle time,
then reaches the other end state.  The doors are either locked or unlocked
depending on the airlock's current state.

The airlock doesn't care or make any decisions based on if people/robots
are inside it or not.  It's just a dumb, simple machine that transitions
when a person/robot activates it.  It's up to the people/robots to use
airlocks properly and safely.  I suppose an analogue to my airlock concept
would be an elevator between two floors in a building.

So the complexity, processes, decision-making, and safety-concerns would be
in the person/robot tasks for passing through the airlock.  I think the
current names of the two related tasks: "ExitAirlock" and "EnterAirlock"
are a bit confusing it just sounds like they're for just entering and
exiting the airlock itself, rather than passing through the airlock to get
into the building/vehicle, or for exiting to the outside.  I'm thinking of
renaming them "EgressThroughAirlock" and "IngressThroughAirlock" and
creating an abstract parent class "AirlockOperation", which will contain
their common code to reduce duplication.

I'm thinking building and vehicle airlocks should be functionally identical
except in occupant capacity.  I imagine a rover airlock might only hold one
or two people at a time, whereas a larger building airlock could hold four
or more people.

Please let me know what you think of this idea.  I want to create something
simple first and see if I can get the tasks working properly with it.  Then
we can start adding more steps and complexity.  We just have to be careful
not to create any bugs along the way.

Scott

**Comment by scud1** [comment]
I may need to merge in new features once I get the simple airlock I'm
working on finished.  I first want to create a reasonably simple airlock
system that doesn't produce any problems with people/robots using it.  No
one gets stuck inside or waiting outside the airlocks, regardless of how
long the simulation is run or the time compression used.

I think the walking code with its tasks and subtasks already handles the
situation of if a person is unable to continue walking to the destination.
Perhaps the person became sick, or not enough oxygen remaining in the EVA
suit, etc.  In this case, the current walking subtask should check for
problems with continuing and end itself early.  The parent waking task
would check for problems and end itself early, etc.

I'll look through the walking code to double-check that it can handle
failures that should cancel the walking tasks.

**Comment by bevans2000** [comment]
@scud1 I like the idea of it being simple and agree that that `EnterAirlock` & `ExitAirlock` can be confusing in terms of names.
In terms of the merging, I can imagine it might be quite difficult. Have you considered creating yourself a branch and committing your changes to that first; giving you a rollback point. Then merge into the branch from master.

**Comment by mokun** [comment]
> I'm still trying to understand your concept of zones and the 4 chambers of an airlock. I see them mentioned in several places but I'm not sure what your overall concept is and how it would work. 

Sorry for the lack of clear descriptions.

Zones are created because the movement through an airlock is unique in each task phase.

Whether it's ingress or egress, there's always a total of 5 zones.

There are 8 task phases in both `EnterAirlock` or `ExitAirlcok`. 

For instance, in `requestEgress()`, you'll see calling method `transitionTo(int zone)`. 

It simply means at some point in that phase, the person will move into that zone.

That's how a zone is connected to the activity of a task phase.

I just added a new section called Zones in the [Airlock wiki](https://github.com/mars-sim/mars-sim/wiki/Airlock). 

Please see below.

------------------------------------

### E. Zones 

Whether it's ingress or egress, there's always a total of 5 zones that a person has to step through.

| # | Zone | 
| --- | ---| 
| 0 | Outside Interior Door | 
| 1 | Inside Interior Door | 
| 2 | In a Chamber | 
| 3 | Inside Exterior Door | 
| 4 | Outside Exterior Door | 

For egress, 
A person first arrives at zone 0, awaiting the interior door to open. 
At zone 1, a person waits to enter into one of the 4 chambers. 
At zone 2, a person dons an EVA suit and fills up with resources.
At zone 3, a person waits for the Exterior Door to open.
At zone 4, a person is officially on the surface of Mars. 

For Ingress, 
A person first arrives at zone 4, awaiting the exterior door to open. 
At zone 3, a person waits to enter into one of the 4 chambers. 
At zone 2, a person doff an EVA suit and empty its resources.
At zone 1, a person waits for the Interior Door to open.
At zone 0, a person leaves the airlock portion of the building and is officially inside a settlement.

**Comment by mokun** [comment]
This [commit](https://github.com/mars-sim/mars-sim/commit/493108bff30b25b8d26f55048093d64be32f59ad) added back the codes in `ExitAirlock`. 

So if the airlock is empty and the chamber is NOT pressurized, anyone who wants to do an EVA egress should be able to pressurize the chamber, unlock the inner door, and walk right in to put on the EVA suit to prepare to go outside.

**Comment by mokun** [comment]
Seems like we've finally managed to clear the bugs that trap the airlock occupants. Let me know if you can run the simulation for a couple hours non-stop without any airlock issues or not. And I'll close this thread in a few days.

**Comment by bevans2000** [comment]
I see People aborting Tasks (not always EVA) but that might just be because they are genuinely tired.

**Comment by mokun** [comment]
> I see People aborting Tasks (not always EVA) but that might just be because they are genuinely tired.

Yes. It's kind of a waste computationally speaking. 

We should not let that task to get started in its meta class in the first place.

**Comment by mokun** [comment]
@bevans2000 ,

Speaking of waste, some of the meta tasks and its associated tasks are performing the same computation. 

So to compute if meta task A is feasible, the cpu will have to do step 1 to 5 to determine if a scientific research or a lab slot is available. 

Finally, this meta task A did get picked.

Now, in the task A, it will redo the same computation to determine the scientific research he will particulate in or a lab slot will be available. 

In that case, we should modify the codes in such a way that it saves the result, just in case the meta task A gets picked. 

If it doesn't get picked, the calculation has been done. No harm done.

But if does get selected, we save some cpu cycles.

**Comment by bevans2000** [comment]
Yes I agree. I had a radical idea when refactoring the Task  but decided it was too far at the time. Instead of the MetaTasks returning a score they return a `WorkPackage` or even a list of them. Each package has the score and a reference to the MetaTask but it also has a set of named value pairs (think of a map keyed on String). The properties might be the details needed to start the equivalent Task identified when creating the score. The appropriate Task would know want to expect and could avoid redoing the work.
For example with Repairs the WorkPackage would contain the Malfunctioning Entity and the Malfunction to fix. In fact it could return more than one if there were multiple malfunctions on the go.
Having named value pairs keeps it nice and flexible, they could contain other object  strings or numbers. Task would be able to start a lot quicker. 
It would be a lot of change since we would not want to have 2 different design active at the same time.

**Comment by bevans2000** [comment]
> > I see People aborting Tasks (not always EVA) but that might just be because they are genuinely tired.
> 
> Yes. It's kind of a waste computationally speaking.
> 
> We should not let that task to get started in its meta class in the first place.

Sometimes a Task is unachievable based on the person executing it which the MetaTask can not predict efficiently. Finding a available airlock for example. It has to cater for Tasks being abandoned at the start once the situation is evaluated. What we don't know easily is why and how many times this is occurring. I feel that the end task method should have a reason strong which we could log. Normal task completion would have no reason and not trigger a log output.

**Comment by mokun** [comment]
> Finding a available airlock for example.

I wonder if taking a tally of how many people who want to *reserve* for an airlock in the same cycle would help or not.

This way, when a person arrive at the airlock, he/she won't be disappointed because the last slot of the airlock has just been taken or something.

**Comment by bevans2000** [comment]
I think this would be difficult because it may take a Person some time to walk to the airlock. Also unexpected things happen in life so the logic has to cater for what is there in the moment.

**Comment by mokun** [comment]
We've seen anyone stuck in an airlock in recent builds anymore. 

If we see it again, we can open a new issue to specifically address that particular situation.
